2018/Vinogradov/Tournament/Tournament/TournamentController.cs

```diff
@@ -0,0 +1,194 @@
+
+                lastPlayerNeedsPlacement = (extraPlayer == 1 && _extraPlayerTour == -1 && Matches[i].Length % 2 == 1);
+
+                if (lastPlayerNeedsPlacement)
+                {
+                    _extraPlayerTour = i + 1;
+                }
+            }
+
+            if (extraPlayer == 1)
+            {
+                Matches[_extraPlayerTour][Matches[_extraPlayerTour].Length - 1].Opponents[1] = numberOfPlayers - 1;
+            }
+
+            Champion = -1;
+            Players = playerNames;
+            Shuffle();
+        }
+
+        private List<int> CountPairs(int numberOfPlayers, int extra)
+        {
+            List<int> matchesOfEachTour = new List<int>();
+            numberOfPlayers -= extra;
+
+            while (extra > 0 || numberOfPlayers > 1)
+            {
+                if (numberOfPlayers % 2 == 1)
+                {
+                    if (extra == 1)
+                    {
+                        numberOfPlayers++;
+                        extra = 0;
+                    }
+                    else
+                    {
+                        numberOfPlayers--;
+                        extra = 1;
+                    }
+                }
+
+                numberOfPlayers = numberOfPlayers / 2;
+                matchesOfEachTour.Add(numberOfPlayers);
+            }
+
+            return matchesOfEachTour;
+        }
+
+        private void Shuffle()
+        {
+            Random rng = new Random();
+
+            for (int i = Players.Length - 1; i > 0; i--)
+            {
+                int j = rng.Next(i + 1);
+
+                if (i != j)
+                {
+                    string tmp = Players[i];
+                    Players[i] = Players[j];
+                    Players[j] = tmp;
+                }
+            }
+        }
+
+        public void PlayGame(MatchInfo match)
+        {
+            if (MatchParametersValidation(match))
+            {
+                if (match.Tour == Matches.Length - 1)
+                {
+                    if (Champion == -1)
+                    {
+                        var currentMatch = Matches[match.Tour][match.MatchNumber];
+                        currentMatch.Winner = match.Winner;
+                        Champion = currentMatch.Opponents[currentMatch.Winner];
+                    }
+
+                    else
+                        throw new InvalidOperationException("This match was already played.");
+                }
+                else
+                {
+                    MovePlayer(match);
+                }
+            }
+            else
+                throw new ArgumentException("Invalid match parameters.");
+        }
+
+        private bool MatchParametersValidation(MatchInfo match)
+        {
+            bool correctTour = (match.Tour >= 0 && match.Tour < Matches.Length);
+            bool correctWinner = (match.Winner == 0 || match.Winner == 1);
+
+            if (correctTour && correctWinner)
+            {
```

Comment:
Shouldn't it be better to do a validation in MatchInfo when creating a new object?